File: app/lib/models/metro_station.dart
import 'package:flutter/material.dart';
import 'package:latlong2/latlong.dart';

/// Metro line classification in Kolkata Metro Rail network
enum KolkataMetroLine {
  blue(
    code: 'blue',
    label: 'Blue Line (Line 1)',
    corridor: 'Dakshineswar ⇄ Kavi Subhash',
    color: Color(0xFF0057B8),
  ),
  green(
    code: 'green',
    label: 'Green Line (Line 2)',
    corridor: 'Howrah Maidan ⇄ Salt Lake Sector-V',
    color: Color(0xFF2E9E4F),
  ),
  purple(
    code: 'purple',
    label: 'Purple Line (Line 3)',
    corridor: 'Joka ⇄ Majerhat',
    color: Color(0xFF7B3FA0),
  ),
  orange(
    code: 'orange',
    label: 'Orange Line (Line 6)',
    corridor: 'Kavi Subhash ⇄ Beleghata',
    color: Color(0xFFF28C1E),
  ),
  yellow(
    code: 'yellow',
    label: 'Yellow Line (Line 4)',
    corridor: 'Noapara ⇄ Jai Hind (Airport)',
    color: Color(0xFFE6B800),
  );

  const KolkataMetroLine({
    required this.code,
    required this.label,
    required this.corridor,
    required this.color,
  });

  final String code;
  final String label;
  final String corridor;
  final Color color;

  static KolkataMetroLine? fromCode(String code) {
    for (final line in KolkataMetroLine.values) {
      if (line.code.toLowerCase() == code.toLowerCase().trim()) return line;
    }
    return null;
  }
}

/// Represents a distinct Kolkata Metro Station entity with precise coordinates
class MetroStation {
  const MetroStation({
    required this.id,
    required this.name,
    required this.line,
    required this.latitude,
    required this.longitude,
    this.nameBn,
    this.code,
    this.layout,
    this.isInterchange = false,
    this.connectingLines = const [],
    this.popularPandalsNearby = const [],
    this.aliases = const [],
    this.opened = const {},
  });

  final String id;
  final String name;
  final String? nameBn;
  final String? code;
  final String? layout;
  final KolkataMetroLine line;
  final double latitude;
  final double longitude;
  final bool isInterchange;
  final List<KolkataMetroLine> connectingLines;
  final List<String> popularPandalsNearby;
  final List<String> aliases;
  final Map<String, String> opened;

  String get displayName => name;
  String get fullDisplayName => (nameBn != null && nameBn!.isNotEmpty) ? '$name ($nameBn)' : name;

  String get subtitle => isInterchange
      ? 'Interchange (${line.label} & ${connectingLines.map((l) => l.label).join(", ")})'
      : line.label;

  LatLng toLatLng() => LatLng(latitude, longitude);

  /// Matches either canonical id (e.g. 'shyambazar') or legacy prefix (e.g. 'm_shyambazar')
  bool matchesId(String queryId) {
    if (id == queryId) return true;
    final cleanQuery = queryId.startsWith('m_') ? queryId.substring(2) : queryId;
    final normalizedId = id.replaceAll('-', '_');
    final normalizedQuery = cleanQuery.replaceAll('-', '_');
    if (normalizedId == normalizedQuery) return true;
    return aliases.any((a) => a.toLowerCase() == queryId.toLowerCase());
  }

  factory MetroStation.fromJson(
    Map<String, dynamic> json, {
    List<String> popularPandals = const [],
    List<String> extraAliases = const [],
  }) {
    final rawLines = (json['lines'] as List<dynamic>?)?.cast<String>() ?? ['blue'];
    final lines = rawLines
        .map((l) => KolkataMetroLine.fromCode(l))
        .whereType<KolkataMetroLine>()
        .toList();

    final primaryLine = lines.isNotEmpty ? lines.first : KolkataMetroLine.blue;
    final connecting = lines.length > 1 ? lines.sublist(1) : <KolkataMetroLine>[];

    final akaList = (json['aka'] as List<dynamic>?)?.cast<String>() ?? [];
    final allAliases = {...akaList, ...extraAliases}.toList();

    final openedMap = <String, String>{};
    if (json['opened'] is Map) {
      (json['opened'] as Map).forEach((k, v) {
        openedMap[k.toString()] = v.toString();
      });
    }

    return MetroStation(
      id: json['id'] as String,
      name: json['name'] as String,
      nameBn: json['nameBn'] as String?,
      code: json['code'] as String?,
      layout: json['layout'] as String?,
      line: primaryLine,
      latitude: (json['lat'] as num).toDouble(),
      longitude: (json['lng'] as num).toDouble(),
      isInterchange: json['interchange'] as bool? ?? (lines.length > 1),
      connectingLines: connecting,
      popularPandalsNearby: popularPandals,
      aliases: allAliases,
      opened: openedMap,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'nameBn': nameBn,
    'code': code,
    'layout': layout,
    'line': line.code,
    'lat': latitude,
    'lng': longitude,
    'isInterchange': isInterchange,
    'connectingLines': connectingLines.map((l) => l.code).toList(),
    'popularPandalsNearby': popularPandalsNearby,
    'aliases': aliases,
    'opened': opened,
  };
}

File: ios/Uma/Models/MetroStation.swift
// MetroStation.swift — port of app/lib/models/metro_station.dart

import Foundation
import SwiftUI
import CoreLocation

enum KolkataMetroLine: String, Codable, CaseIterable, Hashable {
    case blue, green, purple, orange, yellow

    var code: String { rawValue }

    var label: String {
        switch self {
        case .blue:   return "Blue Line (Line 1)"
        case .green:  return "Green Line (Line 2)"
        case .purple: return "Purple Line (Line 3)"
        case .orange: return "Orange Line (Line 6)"
        case .yellow: return "Yellow Line (Line 4)"
        }
    }

    var corridor: String {
        switch self {
        case .blue:   return "Dakshineswar ⇄ Kavi Subhash"
        case .green:  return "Howrah Maidan ⇄ Salt Lake Sector-V"
        case .purple: return "Joka ⇄ Majerhat"
        case .orange: return "Kavi Subhash ⇄ Beleghata"
        case .yellow: return "Noapara ⇄ Jai Hind (Airport)"
        }
    }

    var color: Color {
        switch self {
        case .blue:   return Color(red: 0.0,  green: 0.34, blue: 0.72)
        case .green:  return Color(red: 0.18, green: 0.62, blue: 0.31)
        case .purple: return Color(red: 0.48, green: 0.25, blue: 0.63)
        case .orange: return Color(red: 0.95, green: 0.55, blue: 0.12)
        case .yellow: return Color(red: 0.90, green: 0.72, blue: 0.0)
        }
    }

    static func fromCode(_ code: String?) -> KolkataMetroLine? {
        guard let code = code?.lowercased().trimmingCharacters(in: .whitespaces) else { return nil }
        return KolkataMetroLine(rawValue: code)
    }
}

struct MetroStation: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let name: String
    let nameBn: String?
    let code: String?
    let layout: String?
    let latitude: Double
    let longitude: Double
    let isInterchange: Bool
    let connectingLines: [KolkataMetroLine]
    let popularPandalsNearby: [String]
    let aliases: [String]
    let line: KolkataMetroLine

    enum CodingKeys: String, CodingKey {
        case id, name, nameBn, code, layout
        case latitude = "lat"
        case longitude = "lng"
        case isInterchange = "interchange"
        case connectingLines
        case popularPandalsNearby
        case aliases
        case line
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        nameBn = try? c.decode(String.self, forKey: .nameBn)
        code = try? c.decode(String.self, forKey: .code)
        layout = try? c.decode(String.self, forKey: .layout)
        latitude = try c.decode(Double.self, forKey: .latitude)
        longitude = try c.decode(Double.self, forKey: .longitude)
        let rawLines = (try? c.decode([String].self, forKey: .connectingLines)) ?? []
        let primaryLine: KolkataMetroLine = (try? c.decode(KolkataMetroLine.self, forKey: .line)) ?? .blue
        isInterchange = (try? c.decode(Bool.self, forKey: .isInterchange)) ?? !rawLines.isEmpty
        connectingLines = rawLines.compactMap { KolkataMetroLine(rawValue: $0) }.filter { $0 != primaryLine }
        popularPandalsNearby = (try? c.decode([String].self, forKey: .popularPandalsNearby)) ?? []
        aliases = (try? c.decode([String].self, forKey: .aliases)) ?? []
        line = primaryLine
    }

    init(id: String, name: String, nameBn: String? = nil, code: String? = nil,
         layout: String? = nil, latitude: Double, longitude: Double,
         isInterchange: Bool = false, connectingLines: [KolkataMetroLine] = [],
         popularPandalsNearby: [String] = [], aliases: [String] = [],
         line: KolkataMetroLine) {
        self.id = id; self.name = name; self.nameBn = nameBn; self.code = code
        self.layout = layout; self.latitude = latitude; self.longitude = longitude
        self.isInterchange = isInterchange
        self.connectingLines = connectingLines
        self.popularPandalsNearby = popularPandalsNearby
        self.aliases = aliases; self.line = line
    }

    var coordinate: CLLocationCoordinate2D { .init(latitude: latitude, longitude: longitude) }
    var displayName: String { name }
    var fullDisplayName: String { (nameBn?.isEmpty == false) ? "\(name) (\(nameBn!))" : name }

    var subtitle: String {
        if isInterchange {
            let others = connectingLines.map(\.label).joined(separator: ", ")
            return "Interchange (\(line.label) & \(others))"
        }
        return line.label
    }

    func matchesId(_ queryId: String) -> Bool {
        if id == queryId { return true }
        let cleanQuery = queryId.hasPrefix("m_") ? String(queryId.dropFirst(2)) : queryId
        let normalizedId = id.replacingOccurrences(of: "-", with: "_")
        let normalizedQuery = cleanQuery.replacingOccurrences(of: "-", with: "_")
        if normalizedId == normalizedQuery { return true }
        return aliases.contains { $0.lowercased() == queryId.lowercased() }
    }
}
